File: mindcraft_py/agent_process.py
from __future__ import annotations

import time

from mindcraft_py.node_runtime import NodeRuntimeProcess
from mindcraft_py.python_agent_process import PythonAgentProcess


class AgentProcess(PythonAgentProcess):
    def __init__(self, name, node_runtime=None, logout_agent=None):
        runtime_port = getattr(node_runtime, "port", 0) if node_runtime else 0
        super().__init__(name, runtime_port)
        self.name = name
        self.node_runtime = node_runtime or NodeRuntimeProcess()
        self.logout_agent = logout_agent or (lambda agent_name: None)

    def start(
        self, load_memory=False, init_message=None, count_id=0, profile_path=None
    ):
        process = self.node_runtime.start(
            profiles=[profile_path] if profile_path else None
        )
        super().start(
            load_memory=load_memory,
            init_message=init_message,
            count_id=count_id,
            profile_path=profile_path,
        )
        self.process = process
        return process

    def stop(self):
        self.node_runtime.stop()
        super().stop()

    def handle_exit(self, code=None, signal=None):
        self.logout_agent(self.name)
        if code is not None and code > 1:
            raise SystemExit(code)
        return self.should_restart(code=code, signal=signal)

    def should_restart(self, code=None, signal=None):
        if self._stop_requested or signal == "SIGINT":
            return False
        if code is not None and code > 1:
            return False
        return (time.time() - self.last_restart) >= 10

    def force_restart(self):
        profile_path = None
        if self._start_options:
            profile_path = self._start_options.get("profile_path")
        self.node_runtime.restart(profiles=[profile_path] if profile_path else None)
        return super().force_restart()

    def run(self, profile_path=None, load_memory=False, init_message=None, count_id=0):
        self.start(
            profile_path=profile_path,
            load_memory=load_memory,
            init_message=init_message,
            count_id=count_id,
        )
        return 0

File: src/process/agent_process.ts
import { spawn, ChildProcess } from 'child_process';
import { logoutAgent } from '../mindcraft/mindserver';

export class AgentProcess {
  name: string;
  port: number;
  process: ChildProcess | null = null;
  running: boolean = false;
  count_id: number = 0;

  constructor(name: string, port: number) {
    this.name = name;
    this.port = port;
  }

  start(load_memory: boolean = false, init_message: string | null = null, count_id: number = 0): void {
    this.count_id = count_id;
    this.running = true;

    let args = ['src/process/init_agent.js', this.name];
    args.push('-n', this.name);
    args.push('-c', String(count_id));
    if (load_memory)
      args.push('-l', String(load_memory));
    if (init_message)
      args.push('-m', init_message);
    args.push('-p', String(this.port));

    const agentProcess = spawn('node', args, {
      stdio: 'inherit',
    });

    let last_restart = Date.now();
    agentProcess.on('exit', (code: number, signal: string) => {
      console.log(`Agent process exited with code ${code} and signal ${signal}`);
      this.running = false;
      logoutAgent(this.name);

      if (code > 1) {
        console.log(`Ending task`);
        process.exit(code);
      }

      if (code !== 0 && signal !== 'SIGINT') {
        // agent must run for at least 10 seconds before restarting
        if (Date.now() - last_restart < 10000) {
          console.error(`Agent process exited too quickly and will not be restarted.`);
          return;
        }
        console.log('Restarting agent...');
        this.start(true, 'Agent process restarted.', count_id);
        last_restart = Date.now();
      }
    });

    agentProcess.on('error', (err) => {
      console.error('Agent process error:', err);
    });

    this.process = agentProcess;
  }

  stop() {
    if (!this.running) return;
    this.process?.kill('SIGINT');
  }

  forceRestart() {
    if (this.running && this.process && !this.process.killed) {
      console.log(`Agent process for ${this.name} is still running. Attempting to force restart.`);

      const restartTimeout = setTimeout(() => {
        console.warn(`Agent ${this.name} did not stop in time. It might be stuck.`);
      }, 5000); // 5 seconds to exit

      this.process.once('exit', () => {
        clearTimeout(restartTimeout);
        console.log(`Stopped hanging agent ${this.name}. Now restarting.`);
        this.start(true, 'Agent process restarted.', this.count_id);
      });
      this.stop(); // sends SIGINT
    } else {
      this.start(true, 'Agent process restarted.', this.count_id);
    }
  }
}
